### ISSUE TITLE
Core: improve exceptions and handling

### ISSUE THREAD

**Issue by GPMueller** [spirit-code/spirit#337]
See https://stackoverflow.com/a/348862/4069571 for an example of a custom exception derived from `std::runtime_error`, containing info on both source file and line of code.
The constructor should read `Utility::Spirit_Exception(std::string message, std::string file, unsigned int line).

Part of `Utility::Handle_Exception` should be rewritten:
```C++
catch ( const Utility::Exception & ex )
{
    Handle_Spirit_Exception( ex, idx_image, idx_chain );
}
catch ( const std::exception & ex )
{
    Log( Log_Level::Severe, Log_Sender::API, "Caught std::exception: " + std::string(ex.what()), idx_image, idx_chain );
    Log( Log_Level::Severe, Log_Sender::API, "TERMINATING!", idx_image, idx_chain );
    Log.Append_to_File();

    std::exit( EXIT_FAILURE );  // exit the application. may lead to data loss
}
```
in order to catch the `Spirit_Exception` before `std::exception`, which would include the former.

If we use such a mechanism cleverly, **maybe we can get a hierachical set of messages, telling us exactly the series of function calls that led to the exception being thrown**.
Maybe `Utility::Handle_Exception` should *ONLY* be used int the API, where other functions, when catching an exception, could take the exception message and include it into a new exception which is then thrown.
These should be used as one possible reference: https://stackoverflow.com/a/37227893/4069571 https://stackoverflow.com/a/20028047/4069571

**Comment by GPMueller** [comment]
This answer explains the problem and way to go quite nicely: https://stackoverflow.com/a/8397859/4069571

**Comment by GPMueller** [comment]
Created this MWE for reference and testing: https://github.com/GPMueller/mwe-cpp-exception

**Comment by GPMueller** [comment]
Started working on this feature with 811cce36ed665f8a80a014c8c51d2971629ccde6 on branch `feature-improved-exceptions`. Should be working with fc3557dc16b1300677ce70195eec2fca5e12cc94. Example backtrace:
```
2017-10-26 14:45:07  [SEVERE ] [API ] [--] [--]  ----------------------------------------------------------
2017-10-26 14:45:07  [SEVERE ] [API ] [--] [--]  Exception caught in API function 'State_Setup'
2017-10-26 14:45:07  [SEVERE ] [API ] [--] [--]  SEVERE EXCEPTION
2017-10-26 14:45:07  [SEVERE ] [API ] [--] [--]  Exception backtrace:
2017-10-26 14:45:07  [SEVERE ] [API ] [--] [--]  /Users/gmueller/Git/spirit-test/core/src/io/Configparser.cpp:142 in function 'Spin_System_from_Config': Unable to initialize spin system from config file  "input/input.cfg"
2017-10-26 14:45:07  [SEVERE ] [API ] [--] [--]  /Users/gmueller/Git/spirit-test/core/src/engine/Vectormath.cpp:73 in function 'Build_Spins': Unable to initialize Spin-System, since 2 spins occupy the same space.
Please check the config file!
2017-10-26 14:45:07  [ INFO  ] [ALL ] [--] [--]  Appending Log to file ./Log_2017-10-26_14-45-07.txt
2017-10-26 14:45:07  [SEVERE ] [API ] [--] [--]  TERMINATING!
```

**TODO**:
- maybe add `State *` to `Handle_Exceptions` function to print more detailed info
- improve the `Exception_Classifier`s, rename or hide `S_Exception`, ensure that certain non-breaking exceptions do not force the Handler to terminate execution

**Comment by GPMueller** [comment]
A late realization:
The full backtrace should of course only be done when an exception cannot be recovered, since otherwise even harmless exceptions will cause the respective API function to fail.

I see the following options:
1. catch certain exceptions mid-way and don't propagate them up to the API layer -> call the backtrace in the mid-way function too? Probably this is the best way, the only disadvantage would be that the call tree would not be traced further upwards.
2. make sure exceptions are always severe by covering all recoverable cases some other way -> how to deal e.g. with "file not found"?

**Comment by GPMueller** [comment]
Started reworking according to option 1 with ???. **Need to re-consider every instance of usage of `spirit_rethrow`**: should the exception really be rethrown or should it be handled with `spirit_handle_exception`?

Decisions could look something like this:
```C++
catch( const S_Exception & ex )
{
	Log(Log_Level::Error, Log_Sender::All, "Inside Spin_System_from_Config");
	if (int(ex.level) > 1 && ex.classifier != Exception_Classifier::System_not_Initialized)
		spirit_handle_exception(-1, -1);
	else
		spirit_rethrow( fmt::format("Unable to initialize spin system from config file  \"{}\"", configFile) );
}
```
Unfortunately, I believe, we cannot generalize this `catch`ing, since the exceptions that should be handled are very case-specific. However, maybe something like this might work:
```C++
void special_handle(std::vector<Exception_Classifier> exceptions_to_rethrow, std::string message_if_rethrown)
{
	Log(Log_Level::Error, Log_Sender::All, "Inside Spin_System_from_Config");
	if (int(ex.level) > 1 && !(ex.classifier in exceptions))
		spirit_handle_exception(-1, -1);
	else
		spirit_rethrow( message_if_rethrown );
}
```

**Comment by GPMueller** [comment]
Functionality seems finished with 167a33cdb830114bda418c60af531b3571b2f723.

**The current logic is:**
- `spirit_handle_exception_api` is used to ensure that no exception passes the API layer. It is the final instance and always creates a backtrace. Every API function should contain a `try`/`catch` with this.
- `spirit_handle_exception_core` is used to handle non-severe exceptions below the API layer. It should be used around blocks of code which are not required to pass for the function to work (e.g. a `Geometry` can still be parsed from input, even if the basis file is not found)
- `spirit_rethrow` is used to pass on any exception. This should be used around blocks of code which are required to pass for the respective function to finish.

**TODO:**
- maybe add `State *` to `spirit_handle_exception_api`
- should `spirit_rethrow` be wrapped around all functions which do not use one of the `spirit_handle_exception...` functions?

**Comment by GPMueller** [comment]
Merged into `develop` with 2ef16a4e6b6fe23427f73889838f5717d2ce44c9.
The remaining todo's are now tracked in issue #341.
